[BEAM-1157] Add HBaseIO

iemejia opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [X] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [X] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [X] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

iemejia commented:
R: @dhalperi

iemejia commented:
Rebased to fix error produced by the change of Create.of() => Create.empty

jbonofre commented:
R: @jbonofre

jbonofre commented:
First of all the tests have to be fixed. I gonna take a look.

iemejia commented:
There is something fishy going on, I did mvn clean verify in my machine and everything passes. We need to check if there is something stopping the HBaseTestUtility from been instantiated in jenkins, because the tests that fail are the ones that use the embedded instance.

dhalperi commented:
R: -@jbonofre -- I have already begun a substantial review of this one. Safe your time and enjoy your vacation.

dhalperi commented:
@iemejia I suggest that the precommit bugs are at the least identifying places you can improve error catching. Try adding more error reporting so you can see where things are going wrong.

dhalperi commented on sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java:
```diff
@@ -0,0 +1,681 @@
+        public long getEstimatedSizeBytes(PipelineOptions pipelineOptions) throws Exception {
+            if (estimatedSizeBytes == null) {
+                estimatedSizeBytes = estimateSizeBytes(pipelineOptions);
+                LOG.debug("estimatedSize {} bytes", estimatedSizeBytes);
```
I would also tend to put the `Read` in here.

iemejia replied:
ok

iemejia commented:
Yes sir ! the issue is that almost all the bugs are null pointer exceptions in the objects that should have been initialized in @BeforeClass, I will log initialization to check if this is going well indeed.

dhalperi commented on sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java:
```diff
@@ -0,0 +1,681 @@
+        }
+
+        private List<HRegionLocation> getRegionLocations()
+                throws Exception {
```
inline on previous line?

iemejia replied:
ok

dhalperi commented on sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java:
```diff
@@ -0,0 +1,681 @@
+            // This code is based on RegionSizeCalculator in hbase-server
+            long estimatedSizeBytes = 0L;
+            Configuration configuration = this.read.serializableConfiguration.getConfiguration();
+            try (Connection connection = ConnectionFactory.createConnection(configuration)) {
```
It seems weird that you would create a connection here and then not pass it into getRegionLocations where the connection is then created again.

iemejia replied:
ok

dhalperi commented on sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java:
```diff
@@ -0,0 +1,681 @@
+                    if ((scanWithNoLowerBound
+                            || isLastRegion || Bytes.compareTo(startRow, endKey) < 0)
+                            && (scanWithNoUpperBound || Bytes.compareTo(stopRow, startKey) > 0)) {
+                        regionLocations.add(regionLocation);
```
looks right. Commenting for myself later that I need to look at how well this is tested.

dhalperi replied:
Is there a reason you didn't use the `ByteKey` and `ByteKeyRange` functions like overlap and contains to simplify this logic?

iemejia replied:
At the begining I didn't want to use both ByteKey and ByteKeyRange because I thought they were related to Bigtable not Beam.  I propose you to change this logic afterwards, in a subsequent PR, maybe with the subsplits work, I don't really want to change the current logic.

dhalperi replied:
No, `ByteKey` and `ByteKeyRange` are not Bigtable-specific -- they're specifically added to provide a standard set of common and tricky logic around keys that are `byte[]`.

dhalperi commented on sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java:
```diff
@@ -0,0 +1,681 @@
+                        || Bytes.compareTo(startKey, startRow) >= 0) ? startKey : startRow;
+                final byte[] splitStop =
+                        (scanWithNoUpperBound || Bytes.compareTo(endKey, stopRow) <= 0)
+                                && !isLastRegion ? endKey : stopRow;
```
Is there a reason you didn't use the ByteKey and ByteKeyRange functions like overlap and contains to simplify this logic?

iemejia replied:
See above.

dhalperi commented on sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java:
```diff
@@ -0,0 +1,681 @@
+            LOG.debug("Closing reader after reading {} records.", recordsReturned);
+            if (scanner != null) {
+                scanner.close();
+                scanner = null;
```
will this also close the connection allocated in start?

iemejia replied:
No, excellent catch, fixing right now.

iemejia replied:
ok

dhalperi commented on sdks/java/io/hbase/src/test/java/org/apache/beam/sdk/io/hbase/HBaseIOTest.java:
```diff
@@ -0,0 +1,449 @@
+    @AfterClass
+    public static void afterClass() throws Exception {
+        admin.close();
+        htu.shutdownMiniCluster();
```
these should always look like `if (admin != null) { admin.close(); admin = null; }` that will make your tests more resilient.

iemejia replied:
ok

iemejia replied:
ok

dhalperi commented on sdks/java/io/hbase/src/test/java/org/apache/beam/sdk/io/hbase/HBaseIOTest.java:
```diff
@@ -0,0 +1,449 @@
+            CoderRegistry cr = TestPipeline.create().getCoderRegistry();
+            cr.registerCoder(Mutation.class, HBaseMutationCoder.class);
+        } catch (IOException e) {
+            e.printStackTrace();
```
Why this instead of throw?

iemejia commented on sdks/java/io/hbase/pom.xml:
```diff
@@ -0,0 +1,194 @@
+  <properties>
+    <hbase.version>1.3.0</hbase.version>
+    <hbase.guava.version>12.0.1</hbase.guava.version>
+    <hbase.protobuf.version>2.5.0</hbase.protobuf.version>
```
will do, I doubted, but since this is the current policy I will fix it.

iemejia replied:
ok

iemejia commented on sdks/java/io/hbase/pom.xml:
```diff
@@ -0,0 +1,194 @@
+      <artifactId>commons-lang</artifactId>
+      <version>2.6</version>
+      <scope>test</scope>
+    </dependency>
```
This one is scoped to test so it should not leak. I don't think we need to shade it. Or do you have a strong reason to do it ?

dhalperi replied:
Ack, you're right, I missed the `test`

iemejia commented on sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java:
```diff
@@ -0,0 +1,681 @@
+ *         .withTableId("table"))
+ *         .withScan(scan));
+ *
+ * // Scan a prefix of the table.
```
I copy exactly this lined from the BigtableIO doc, tell me what you prefer and I will change it.
https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java#L102

dhalperi replied:
+1

iemejia commented on sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java:
```diff
@@ -0,0 +1,681 @@
+ *
+ * <h3>Experimental</h3>
+ *
+ * <p>The design of the API for HBaseIO is currently related to the BigTableIO one,
```
ok

iemejia commented on sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java:
```diff
@@ -0,0 +1,681 @@
+
+        @Override
+        public PCollection<Result> expand(PBegin input) {
+            HBaseSource source = new HBaseSource(this, null);
```
ok, again idea copied from BigtableSource that does not do this. I will fix as you proposed.

iemejia commented on sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java:
```diff
@@ -0,0 +1,681 @@
+        @Override
+        public void validate(PBegin input) {
+            checkArgument(serializableConfiguration != null,
+                    "Configuration not provided");
```
good catch.

iemejia commented on sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java:
```diff
@@ -0,0 +1,681 @@
+                                serializableConfiguration.getConfiguration());
+                Admin admin = connection.getAdmin();
+                checkArgument(admin.tableExists(TableName.valueOf(tableId)),
+                        "Table %s does not exist", tableId);
```
oh yes silly mistake thanks for noticing.

iemejia replied:
ok

iemejia commented on sdks/java/io/hbase/pom.xml:
```diff
@@ -0,0 +1,194 @@
+    <dependency>
+      <groupId>org.apache.beam</groupId>
+      <artifactId>beam-runners-direct-java</artifactId>
+      <version>${project.version}</version>
```
Not really, every other IO (again including google ones) includes it like this, if I remove this line I get an "IllegalArgument No Runner was specified"

iemejia replied:
ok

iemejia commented on sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java:
```diff
@@ -0,0 +1,681 @@
+ * <pre>{@code
+ * Configuration configuration = ...
+ * PCollection<KV<ByteString, Iterable<Mutation>>> data = ...;
+ * data.setCoder(HBaseIO.WRITE_CODER)
```
Best for the end, this one I really need you to give me some hints.
Notice that to make the API similar to Bigtable, I let this Mutation there, however HBase mutation is not Serializable. So I need to encode it to pass it, but I didn't find how to do it, because if the user is doing it externally he can use a CoderRegistry or do the setCoder in the PCollection.
I don't know how to solve it otherwise because I would need something like the opposite of getDefaultOutputCoder. I mean, like getDefaultInputCoder, and I didn't find this in the API or is there any other way to do this that I am missing, can you give me some ideas ?

dhalperi replied:
Ha! I literally meant "there's a missing `;`".

@kennknowles filed a JIRA recently about registering coders dynamically. As is, something like this may be required, or of course the user adds custom configuration to the pipeline coder registry.

iemejia replied:
Hehe ok, can you please refer me to the JIRA number to start watching it and improve this once it is solved.

iemejia commented on sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java:
```diff
@@ -0,0 +1,681 @@
+         * <p>Does not modify this object.
+         */
+        public Read withKeyRange(byte[] startRow, byte[] stopRow) {
+            checkNotNull(startRow, "startRow");
```
ok, I did this by making those into a KeyRange and making the scan from the KeyRange now

dhalperi commented:
This looks to me like HBase is trying to bind an internet-public IP for
listening, which is likely not allowed. It should be binding something like
`localhost`. Is this something you can override in the test config?

On Mon, Feb 13, 2017 at 9:34 AM, asfbot <notifications@github.com> wrote:

> Refer to this link for build results (access rights to CI server needed):
> https://builds.apache.org/job/beam_PreCommit_Java_MavenInstall/7357/
> Failed Tests: 1beam_PreCommit_Java_MavenInstall/org.apache.beam:
> beam-sdks-java-io-hbase
> <https://builds.apache.org/job/beam_PreCommit_Java_MavenInstall/7357/org.apache.beam$beam-sdks-java-io-hbase/testReport>:
> 1
>
>    -
>    *org.apache.beam.sdk.io.hbase.HBaseIOTest.org.apache.beam.sdk.io.hbase.HBaseIOTest*
>    <https://builds.apache.org/job/beam_PreCommit_Java_MavenInstall/7357/org.apache.beam$beam-sdks-java-io-hbase/testReport/org.apache.beam.sdk.io.hbase/HBaseIOTest/org_apache_beam_sdk_io_hbase_HBaseIOTest/>
>
>
> --none--
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/beam/pull/1961#issuecomment-279463195>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAgIT_1yzmfpWs87Kc_PB3f20u5y6OEgks5rcJQvgaJpZM4L8Sg9>
> .
>

iemejia commented:
Arrghh I was lucky to be close to 1 test to pass :)
I am going to try something as you said, this one looks similar, so probably the variable will make it.
https://issues.apache.org/jira/browse/HBASE-11711

dhalperi commented on sdks/java/io/hbase/src/test/java/org/apache/beam/sdk/io/hbase/HBaseIOTest.java:
```diff
@@ -99,21 +99,21 @@
     public static void beforeClass() throws Exception {
-        LOG.info("Starting HBase Embedded Server (HBaseTestUtility)");
+        System.out.println("Starting HBase Embedded Server (HBaseTestUtility)");
         conf.setInt(HConstants.HBASE_CLIENT_RETRIES_NUMBER, 1);
```
I think that ultimately what you want is to, in your configuration, set the HBase server name to localhost or 127.0.0.1

jbonofre commented:
Now that I'm back from vacation, I will help for the review with Dan.

dhalperi commented:
I took a pass on the HBase Configuration in #2036. The changes I made to the test seem to have worked -- the most recent Jenkins run number [7544](https://builds.apache.org/job/beam_PreCommit_Java_MavenInstall/7544/org.apache.beam$beam-sdks-java-io-hbase/testReport/org.apache.beam.sdk.io.hbase/HBaseIOTest/) actually tested and passed HBaseIOTest. [Jenkins build failed overall because of the awful hacks I did to the poms to streamline testing.]

Please take a look -- I hope this unblocks you getting this green!

dhalperi commented:
https://github.com/apache/beam/pull/2036/files/cbf29331a53102680fcbd468757f8521c8292cff..8f2be281b84d3d47e5d7db8bf7c1f8d20013e209 was my specific hacks. (ignore anything outside of HBaseIOTest)

iemejia commented:
Finally it passed !
Thanks a lot than for the help to fix it, and for the review. Now it is up to you to tell me if I need to fix something else or we are ready to go.

dhalperi commented on sdks/java/io/hbase/pom.xml:
```diff
@@ -0,0 +1,238 @@
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-jar-plugin</artifactId>
+      </plugin>
```
I think most of these should be inherited. Are they needed?

iemejia replied:
Yes, they should be put explicitly, all the other IOs do the same.

dhalperi replied:
I do not think they are needed. Please look at https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/pom.xml

The only place we have them is in `pluginManagement`, to override config from parent.

iemejia replied:
https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/pom.xml#L41 ?

iemejia replied:
I have prepared a subsequent PR to remove the need to import these three maven plugins from all the IOs, but I am hoping for this one to get merged to open it.

dhalperi replied:
Hmm, clearly I'm crazy. Thank you.

dhalperi commented on sdks/java/io/hbase/src/test/java/org/apache/beam/sdk/io/hbase/HBaseIOTest.java:
```diff
@@ -0,0 +1,439 @@
+        for (Result result : scanner) {
+            results.add(result);
+        }
+        System.out.println(estimatedBytesSize);
```
drop `println` throughout.

dhalperi commented on sdks/java/io/hbase/src/test/java/org/apache/beam/sdk/io/hbase/HBaseIOTest.java:
```diff
@@ -0,0 +1,439 @@
+        htu = new HBaseTestingUtility(conf);
+        htu.startMiniCluster(1, 4);
+        admin = htu.getHBaseAdmin();
+        System.out.println("Started HBase Embedded Server (HBaseTestUtility)");
```
drop println.

dhalperi replied:
(LOG okay)

iemejia replied:
Oh that was a mistake from trying to do jenkins happy, thanks for noticing.

iemejia replied:
ok

dhalperi commented on sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/coders/SerializableScan.java:
```diff
@@ -0,0 +1,49 @@
+/**
+ * This is just a wrapper class to serialize HBase {@link Scan}.
+ */
+public class SerializableScan implements Serializable {
```
Are these intended for use outside of the `hbase` module, or only inside of the HBaseIO module? If the latter, I would not make them `public`.

iemejia replied:
They are not, however as they live in the coders package, I have to make them public, otherwise I cannot import them. One approach would be to move them to the same directory but I think this makes the code more messier (even if I agree with you about the desired access level).

iemejia commented:
Ok, fixed last remarks, rebased into one single commit and rebased, I hope it is ok right now so it can be merged soon.
Thanks again for the review and the patience Dan.

iemejia commented:
Notice that this extra rebase was to add the IO to the parent pom and the javadoc pom.

jbonofre commented:
I already updated the other IOs. So, please, don't update the other IO in another PR.

dhalperi commented:
Merged! Can you file JIRAs for follow-on work (adding support for Dynamic work rebalancing, improving splitting, switching to ByteKey, ByteKeyRange, ByteKeyRangeTracker to help the first 2)?

iemejia commented:
Awesome, thanks a lot Dan for your review + the binding fix, I will be filling the JIRAs now.
Thanks a lot also JB for your feedback.

iemejia commented:
JB, as discussed, I will do the update to remove extra plugins for the other IOs, thanks for letting me that one.

jbonofre commented:
Sure ! Thanks ! I set my branch "on hold" ;)
